Feature/api basic auth

OwenTuz opened the pull request:
#10615 - Add basic auth to REST API

```
Adds @requires_auth decorator to API, which checks basic auth against a SHA512-hashed password
Adds CLI options for 'username' and 'password'
Adds YAML configuration option for API (using CLI option "-c /path/to/file.yaml")
Updates API tests, Docker API script and documentation to reflect the change
```

Note that this breaks backwards compatibility by forcing a requirement for authentication on all views (ie users must specify a password hash either on the CLI or in a configuration file - there's no default). It also requires users running the Docker script to specify a configuration file containing at minimum their hashed password.

I'm happy to refactor that if we would prefer to make basic authentication optional and retain backward compatibility.

OwenTuz committed:
Add HTTP basic auth decorator @requires_auth to API

OwenTuz committed:
Add username (optional) and password (required) to API invocation arguments

OwenTuz committed:
Require auth for all API views

Commit:
Require basic auth for API unit tests

Commit:
Add option to specify a YAML file containing username/password

Commit:
Allow users to specify "host" and "port" in YAML config

Commit:
Fix misplaced lower() method for better comparison

Commit:
Move "verbose" under server options

Commit:
Require API configuration file when using Docker script

Commit:
Add documentation for authentication and config file usage

Commit:
Add API SSL warning

andresriancho commented on doc/sphinx/api/index.rst:
```diff
@@ -23,9 +23,70 @@ Or it can also be run inside a docker container:
+running:
+
+.. code-block:: none
+    $ echo -n "secret" | sha512sum
```
Add empty line after the `code-block` directive

andresriancho commented on doc/sphinx/api/index.rst:
```diff
@@ -23,9 +23,70 @@ Or it can also be run inside a docker container:
+    $ echo -n "secret" | sha512sum
+    bd2b1aaf7ef4f09be9f52ce2d8d599674d81aa9d6a4421696dc4d93dd0619d682ce56b4d64a9ef097761ced99e0f67265b5f76085e5b0ee7ca4696b2ad6fe2b2  -
+
+If the above hash was specified using the "password" option, users would be able 
```
Replace was with is

andresriancho commented on doc/sphinx/api/index.rst:
```diff
@@ -23,9 +23,70 @@ Or it can also be run inside a docker container:
+For example, using the 'curl' command:
+
+.. code-block:: none
+    $ curl -u admin:secret http://127.0.0.1:5000
```
Add empty line after `code-block`

andresriancho commented on doc/sphinx/api/index.rst:
```diff
@@ -23,9 +23,70 @@ Or it can also be run inside a docker container:
+      "docs": "http://docs.w3af.org/en/latest/api/index.html" 
+    }
+
+Although this provides some security, traffic passing to and from the API is not
```
> Please note that even with basic authentication, traffic passing to and from the API is not

andresriancho commented on doc/sphinx/api/index.rst:
```diff
@@ -23,9 +23,70 @@ Or it can also be run inside a docker container:
+Config file format
+------------------
+
+Using a configuration file is completely optional unless you're starting the
```
Remove "completely"

andresriancho commented on extras/docker/scripts/w3af_api_docker:
```diff
@@ -15,6 +15,9 @@ def parse_args():
                                             ' the latest w3af features.',
                         required=False)
+    parser.add_argument('-c', '--config-file',
+                        help='[Required] Path to API config file in YAML format.',
```
Remove the `[Required]`, argparse will show the required arguments in a different way to the user so this text is redundant

andresriancho commented on w3af/core/ui/api/main.py:
```diff
@@ -51,17 +52,47 @@ def parse_arguments():
+                        default=False,
+                        nargs='?')
+
+    parser.add_argument('-c',
```
I believe that argparse provides a "type" for input files, if you specify that argparse will make sure the file exists, is readable, etc.

andresriancho commented on w3af/core/ui/api/main.py:
```diff
@@ -86,16 +121,55 @@ def main():
+        return 1
 
-    if args.host not in ('127.0.0.1', 'localhost'):
-        print('The REST API does not provide authentication and might expose'
```
Maybe replace this message with a warning about SSL? (like in the docs)

andresriancho commented on w3af/core/ui/api/main.py:
```diff
@@ -70,10 +101,14 @@ def parse_arguments():
-    args.port = host_port[1]
-
+    try:
+      args.host, args.port = getattr(args,'host:port').split(':')
```
Please use 4-spaces for indent (this applies to all the code)

andresriancho commented on w3af/core/ui/api/main.py:
```diff
@@ -86,16 +121,55 @@ def main():
+
+    if (app.config['HOST'] == '127.0.0.1' or
+        app.config['HOST'] == 'localhost'):
+      print('CAUTION! Traffic to this API is not encrypted and could be sniffed.'
```
Ah! You did add this :+1:

andresriancho commented:
I like your PR and do believe that we should let the user start the API without authentication (default). When starting without auth we should print a warning with the two issues (no auth, no ssl).

Commit:
Fix indentation

Commit:
Remove unnecessary [Required] from help text

Commit:
Use argparse FileType for built-in checking

Commit:
Spacing; Fix check that did exactly the opposite of what it was supposed to do

Commit:
Add type check (avoids passing an open file object into app.config)

Commit:
Make basic auth optional (preserve backward compatibility); reinstate warning message if running on public IP without auth

Commit:
Close config file after use since argparse.FileType does not

Commit:
Remove requirement for config file from Docker script, make "debug" and "config file" mutually exclusive

Commit:
Update documentation

Commit:
Quote password when starting API for tests

OwenTuz commented:
Thanks for taking a look over it! I've updated this based on the above.

- Various code and documentation formatting fixes
- Authentication is now optional and only applied if a password is specified on the CLI or in a config file
- Unit tests still use authentication
- Fixed a daft bug where I had specified '==' instead of '!=' (...)
- Switched to using argparse.FileType to test file existence.
- Tightened up type checking when copying key/value pairs into Flask's app.config
- Made 'debug' and 'config-file' options mutually exclusive arguments in the Docker script (since 'debug' runs bash and bypasses the API command line entirely)

Commit:
Fix broken import in unit tests

andresriancho commented:
It all looks good now, anything else that's pending? If it looks good to you, I'll git pull your code, test it live and merge if all works

OwenTuz commented:
I think that's the lot - please do. You know where to find me when you spot the bugs.

andresriancho commented:
Ok - will review this tomorrow

andresriancho commented:
Take a look at https://github.com/andresriancho/w3af/commit/4e1a4aa482b4f4dc62128a20ba6eb7089d04c855 and https://github.com/andresriancho/w3af/commit/b18018f31f5f96adc10282fcee9099a1169ebb59 for a couple of improvements on your already great work :+1:

andresriancho commented:
Tests are running at https://circleci.com/gh/andresriancho/w3af/2140 , after they PASS I'll merge into develop

andresriancho commented:
Build success, merging! Thanks for the contribution, keep them coming, some suggestions:
 * https://github.com/andresriancho/w3af/issues/10831
 * https://github.com/andresriancho/w3af/issues/10616

OwenTuz commented:
Thanks :) Will take a look.
